docker-app/qfieldcloud/core/permissions_utils.py

```diff
@@ -271,6 +271,12 @@ def can_create_delta(user: QfcUser, delta: Delta) -> bool:
     """Whether the user can store given delta."""
     project: Project = delta.project
 
+    if (
+        project._has_online_vector_data
+        and not project.owner.useraccount.account_type.is_external_db_supported
+    ):
+        return False
+
```

Comment:
The only potential problem I see here is that the UI might use this permission check.

Reply:
Yes indeed... In the case of the similar feature about checking quota when changing owner, we ended up doing the check separately both in the API view and in the GUI view (so not ideal in terms of not repeating oneself)...

I'll see if I can implement this in the Delta's `clean()` method ([docs](https://docs.djangoproject.com/en/4.0/ref/models/instances/#validating-objects-1)).

About the second part of your comment (whether we block all deltas, or only the once about these layers) it would be possible to implement this I guess. But not sure what's best ? IMO better to have no delta apply rather than just some, because that could break the data in case there are some relations. But not sure how it's done for other type of failures ?

Let me know what you prefer.

Reply:
Hmm actually `clean()` is not so well adapted, as it is supposed to be for hard validation (you can't save unless valid). Anyway, I'll just move that to a property on the model, so we can reuse it if needed (I think it's better on the model itself, but arguably we could also put in to permission_utils, or even the other way around on the account_type).

Then I guess it would be good for the user to know why the delta is failing (because that's not obvious). With current implementation he would just see that it's UNPERMITTED, but not why. In the frontend GUI, we could just call the same function to display ad hoc message. But how to deal with that through the API ? Is it needed ? Should we somehow store the reason it is UNPERMITTED ? Can we reuse `last_feedback`, or should we introduce a specific field ?

That makes me think that the initial approach (trying to apply the delta, and failing once we try) was not too bad, as this way we'd just tie into the existing delta failure implementation.

Let me know what you think (maybe we need another call to clarify exactly what we want, including all user facing aspects)

Reply:
> Then I guess it would be good for the user to know why the delta is failing (because that's not obvious). With current implementation he would just see that it's UNPERMITTED, but not why. In the frontend GUI, we could just call the same function to display ad hoc message. But how to deal with that through the API ? Is it needed ? Should we somehow store the reason it is UNPERMITTED ? Can we reuse last_feedback, or should we introduce a specific field ?

I think it would be more than enough to:
1) reply with an error when uploading deltas, that they do not match your current subscription plan
2) when opening the deltas page in the browser, show "You cannot apply deltas because your subscription plan"

 > That makes me think that the initial approach (trying to apply the delta, and failing once we try) was not too bad, as this way we'd just tie into the existing delta failure implementation.
This will bring extra jobs, which is the heaviest part in the whole QFC stack and would be nice to avoid.

> About the second part of your comment (whether we block all deltas, or only the once about these layers) it would be possible to implement this I guess. But not sure what's best ? IMO better to have no delta apply rather than just some, because that could break the data in case there are some relations. But not sure how it's done for other type of failures ?

Probably would be easier to describe that no deltas can be applied if you are not having a good subscription plan, agreed.